reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,135 @@ object StreamSubscriber {
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription =>
+            F.delay(s.request(maxDemand)).as(WaitingOnUpstream(s))
+          case Uninitialized =>
+            (Idle(s): State).pure[F]
+          case Idle(_) =>
+            F.delay(s.cancel).as(Idle(s))
           case o =>
             val err = new Error(s"received subscription in invalid state [$o]")
-            o -> (F.delay(s.cancel) >> F.raiseError(err))
+            (F.delay(s.cancel) >> F.raiseError(err)).as(o)
         }
         case OnNext(a) => {
-          case WaitingOnUpstream(s, r) => Idle(s) -> r.complete(a.some.asRight)
-          case DownstreamCancellation  => DownstreamCancellation -> F.unit
-          case o                       => o -> F.raiseError(new Error(s"received record [$a] in invalid state [$o]"))
+          case WaitingOnUpstream(s) =>
+            q.enqueue1(a.some.asRight).as(Receiving(s))
+          case Receiving(s) =>
+            q.enqueue1(a.some.asRight).as(Receiving(s))
+          case Idle(s) =>
+            q.enqueue1(a.some.asRight).as(Receiving(s))
+          case DownstreamCancellation =>
+            (DownstreamCancellation: State).pure[F]
+          case o =>
+            F.raiseError(new Error(s"received record [$a] in invalid state [$o]")).as(o)
         }
         case OnComplete => {
-          case WaitingOnUpstream(_, r) => UpstreamCompletion -> r.complete(None.asRight)
-          case _                       => UpstreamCompletion -> F.unit
+          case WaitingOnUpstream(_) =>
+            q.enqueue1(None.asRight).as(UpstreamCompletion)
+          case Receiving(_) =>
+            q.enqueue1(None.asRight).as(UpstreamCompletion)
+          case Idle(_) =>
+            q.enqueue1(None.asRight).as(UpstreamCompletion) // empty queues will complete while idle
+          case _ =>
+            (UpstreamCompletion: State).pure[F]
         }
         case OnError(e) => {
-          case WaitingOnUpstream(_, r) => UpstreamError(e) -> r.complete(e.asLeft)
-          case _                       => UpstreamError(e) -> F.unit
+          case WaitingOnUpstream(_) =>
+            (q.enqueue1(e.asLeft) >> q.enqueue1(None.asRight)).as(UpstreamError(e))
+          case Receiving(_) =>
+            q.enqueue1(None.asRight).as(UpstreamError(e))
+          case _ =>
+            (UpstreamError(e): State).pure[F]
         }
         case OnFinalize => {
-          case WaitingOnUpstream(sub, r) =>
-            DownstreamCancellation -> (F.delay(sub.cancel) >> r.complete(None.asRight))
-          case Idle(sub) => DownstreamCancellation -> F.delay(sub.cancel)
-          case o         => o -> F.unit
+          case WaitingOnUpstream(sub) =>
+            (F.delay(sub.cancel) >> q.enqueue1(None.asRight)).as(DownstreamCancellation)
```

Comment:
It seems like the handling in all these cases should be the same, `OnFinalize` only happens when the stream terminates, so we know nothing further will be dequeued.